File: Java/src/com/leetcode/Medium/FindLargestValueInEachTreeRow.java
package com.leetcode.Medium;

import java.util.*;

// 515. Find Largest Value in Each Tree Row
public class FindLargestValueInEachTreeRow {
    public static class TreeNode {
        int val;
        TreeNode left, right;
    }

    public List<Integer> largestValues(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null)
            return res;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            int max = Integer.MIN_VALUE;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                max = Math.max(max, node.val);

                if (node.left != null)
                    queue.add(node.left);

                if (node.right != null)
                    queue.add(node.right);
            }
            res.add(max);
        }
        return res;
    }
}

File: JavaScript/Medium/FindLargestValueInEachTreeRow.js
/**
 * 515. Find Largest Value in Each Tree Row
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
const largestValues = function (root) {
  let res = [];
  if (root == null) {
    return res;
  }

  let queue = [];
  queue.push(root);
  while (queue.length !== 0) {
    let max = Number.MIN_SAFE_INTEGER;
    let size = queue.length;

    for (let i = 0; i < size; i++) {
      let node = queue.shift();
      max = Math.max(max, node.val);

      if (node.left !== null) {
        queue.push(node.left);
      }

      if (node.right !== null) {
        queue.push(node.right);
      }
    }

    res.push(max);
  }
  return res;
};